src/main/scala/analysis/DSAUtility.scala

```diff
@@ -0,0 +1,975 @@
+      cells.update(offset, cell)
+      cell.growSize(size)
+      cell
+    else
+      cells(offset).growSize(size)
+      cells(offset)
+
+  def cloneSelf(graph: DSG) : DSN =
+    val node = DSN(Some(graph), this.size)
+    node.allocationRegions.addAll(this.allocationRegions)
+    node.flags.join(this.flags)
+    cells.foreach{
+      case (offset: BigInt, cell: DSC) =>
+        node.addCell(offset, cell.largestAccessedSize)
+    }
+    node
+
+  def cloneNode(from: DSG, to: DSG): Unit =
+//    assert(from.nodes.contains(this)) TODO update nodes after each phase for to check this assertion
+    if !to.nodes.contains(this) then
+      to.nodes.add(this)
+
+
+      from.varToCell.foreach(
+        t =>
+          val pos = t._1
+          val varMap = t._2
+          varMap.foreach{
+            case (variable: Variable, slice: Slice) =>
+              if from.find(slice).node.equals(this) then
+                to.varToCell.update(
+                  pos,
+
+                  to.varToCell.getOrElseUpdate(pos,
+                    mutable.Map[Variable, Slice]()) ++ Map(variable -> from.find(slice))
+                )
+          }
+      )
+      from.formals.foreach{
+        case (variable: Variable, slice: Slice) =>
+          if from.find(slice).node.equals(this) then
+            to.varToCell.update(
+              begin(from.proc),
+              to.varToCell.getOrElseUpdate(begin(from.proc),
+                mutable.Map[Variable, Slice]()) ++ Map(variable -> from.find(slice))
+            )
+      }
+
+      cells.foreach {
+        case (offset: BigInt, cell: DSC) =>
+        if cell._pointee.isDefined then
+          val pointee = cell.getPointee
+          pointee.node.cloneNode(from, to)
+//          to.pointTo.update(cell, pointee) TODO check this is not necessary
+      }
+
+  override def equals(obj: Any): Boolean =
+    obj match
+      case node: DSN =>
+        this.id == node.id
+      case _ => false
+
+  override def hashCode(): Int = id
+
+  override def toString: String = s"Node($id, $allocationRegions ${if collapsed then ", collapsed" else ""})"
+
+
+}
+
+/**
+ * a cell in DSA
+ * @param node the node this cell belongs to
+ * @param offset the offset of the cell
+ */
+case class DSC(node: Option[DSN], offset: BigInt)
+{
+  var largestAccessedSize: BigInt = 0
+
+  var _pointee : Option[Slice] = None
+
+  def getPointee : Slice =
+    if _pointee.isEmpty then
+      val node = DSN(Some(this.node.get.graph.get))
+      _pointee = Some(Slice(node.cells(0), 0))
+    else
+
+      val graph = _pointee.get.node.graph.get
+      val resolvedPointee = graph.find(graph.adjust(_pointee.get))
+
+      _pointee = Some(graph.deadjust(resolvedPointee))
+    _pointee.get
+
+  def growSize(size: BigInt): Boolean =
+    if size > largestAccessedSize then
+      largestAccessedSize = size
+      true
+    else false
+
+  override def toString: String = s"Cell(${if node.isDefined then node.get.toString else "NONE"}, $offset)"
+}
```

Comment:
This should be a class, not a case class, as it contains mutable variables.

Is the idea that `_pointee` is private? You have not made it private.

Reply:
Sometimes I want to check if _pointee is defined or not and othertimes I want to get a cell's pointee or create one for it if it doesn't have one. getPointee is not meant as an accessor  to a private variable but a helper that checks if a pointee is already defined and if not creates  one. I still want to read and write to _pointee outside of DSC class. I can make it private and create accessor methods for it but I don't really want to restrict other classes from accessing _pointee

Reply:
This is a bit unclear in the current code and better names or comments would probably help.

What's with the _ in the name then? In some languages that's used for private variables in a class?

Reply:
you are right. I'll rename it